### PR TITLE
Manual partition assignment

### PR DESCRIPTION
**Pull request by terciodemelo**
Automatic random generation of partition assignments limits the user control over them. At times it is necessary, or at least desirable to tweak the assignment and manually make decisions on which replica of which partition goes to. An example of such a situation is when only a small subset of topics takes most of data flow on the cluster, and you may wanna make sure they are separated, in different brokers.

A new view was created for these decisions, where the user can see both the current assignment and some broker metrics to aid on choices. Two metrics were also added to this view: Process CPU Load and System CPU Load.
screen shot 2015-07-06 at 11 06 44 pm

This view can be accessed through the route /clusters/:cluster/assignments/manual or by clicking the button "Manually Set Partition Assignmentns" in /clusters/:cluster/topics.

I hope more people find it useful!

EDIT 1: I also added two filters in this view: to filter topics and to filter metrics, in order to decrease the visual clutter on the page.

EDIT 2: Feedbacks are welcome =p

### CODE REVIEW

**Comment by patelh** [comment]
Looks cool.  Will have to try it out and will let you know.

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 1]
```diff
@@ -65,6 +65,38 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
     log.error("Long running pool queue full, skipping!")
   }
 
+  private def allBrokerViews(): Seq[BVView] = {
+    var bvs = mutable.MutableList[BVView]()
+    for (key <- brokerTopicPartitions.keySet.toSeq.sorted) {
+      val bv = brokerTopicPartitions.get(key).map { bv =>
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Can we move the logic used here and BVGetView into a method to reduce duplication.  Maybe a new method like 
```
produceBViewWithBrokerClusterState(bv: BVView) : BVView = {
val bcs = for {
            metrics <- bv.metrics
            cbm <- combinedBrokerMetric
          } yield {
            val perMessages = if(cbm.messagesInPerSec.oneMinuteRate > 0) {
              BigDecimal(metrics.messagesInPerSec.oneMinuteRate / cbm.messagesInPerSec.oneMinuteRate * 100D).setScale(3, BigDecimal.RoundingMode.HALF_UP)
            } else ZERO
            val perIncoming = if(cbm.bytesInPerSec.oneMinuteRate > 0) {
              BigDecimal(metrics.bytesInPerSec.oneMinuteRate / cbm.bytesInPerSec.oneMinuteRate * 100D).setScale(3, BigDecimal.RoundingMode.HALF_UP)
            } else ZERO
            val perOutgoing = if(cbm.bytesOutPerSec.oneMinuteRate > 0) {
              BigDecimal(metrics.bytesOutPerSec.oneMinuteRate / cbm.bytesOutPerSec.oneMinuteRate * 100D).setScale(3, BigDecimal.RoundingMode.HALF_UP)
            } else ZERO
            BrokerClusterStats(perMessages, perIncoming, perOutgoing)
          }
          if(bcs.isDefined) {
            bv.copy(stats=bcs)
          } else {
            bv
          }
}
```

Comment (terciodemelo) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Sure! I'll do it right now

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 2]
```diff
@@ -143,6 +179,7 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
   }
 
   private[this] def updateView(): Unit = {
+    var nBrokers: Int = 0
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: This is not needed.  You can fix the missing brokers from the list by modifying lines 157 - 161.  You can keep an empty broker view:
```
val EMPTY_BVVIEW = BVView(Map.empty, config.clusterConfig, Option(BrokerMetrics.DEFAULT)).
```

And then change the following on line 157 :
```
        for {
          bv <- brokerTopicPartitions.get(id)
        } {
          brokerTopicPartitions.put(id, bv.copy(metrics = Option(metrics)))
        }
```

To something like:
```
          val updatedBVView = brokerTopicPartitions.getOrElse(id, EMPTY_BVVIEW).copy(metrics = Option(metrics))
          brokerTopicPartitions.put(id, updatedBVView)
 ```

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 3]
```diff
@@ -187,6 +224,8 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
             }
         }
 
+        nBrokers = brokerList.list.size
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Not needed, see comment above.

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 4]
```diff
@@ -209,9 +248,12 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
       } else if(config.clusterConfig.jmxEnabled) {
         log.warning("Not scheduling update of JMX for all brokers, not enough capacity!")
       }
-      
+
+      var brokersWithTopics = mutable.MutableList[Int]()
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Not needed, see comment above.

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 5]
```diff
@@ -209,9 +248,12 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
...
+
       topicPartitionByBroker.foreach {
         case (brokerId, topicPartitions) =>
+          brokersWithTopics += brokerId
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Not needed, see comment above.

File: app/kafka/manager/BrokerViewCacheActor.scala [hunk 6]
```diff
@@ -219,6 +261,13 @@ class BrokerViewCacheActor(config: BrokerViewCacheActorConfig) extends LongRunni
           brokerTopicPartitions.put(
             brokerId,BVView(topicPartitionsMap, config.clusterConfig, brokerMetrics.get(brokerId)))
       }
+      for (i <- 0 until nBrokers) {
```
Comment (patelh) [inline app/kafka/manager/BrokerViewCacheActor.scala]: Not needed, see comment above.

File: app/kafka/manager/ClusterManagerActor.scala [hunk 7]
```diff
@@ -294,6 +295,27 @@ class ClusterManagerActor(cmConfig: ClusterManagerActorConfig)
         }
         result.map(CMCommandResults.apply) pipeTo sender()
 
+      case CMManualPartitionAssignments(assignments) =>
+        implicit val ec = longRunningExecutionContext
+        val result = Future {
+          modify {
```
Comment (patelh) [inline app/kafka/manager/ClusterManagerActor.scala]: This looks to be duplicating 279-295, can we move to a method so we can reuse in both places.

File: app/kafka/manager/ClusterManagerActor.scala [hunk 8]
```diff
@@ -294,6 +295,27 @@ class ClusterManagerActor(cmConfig: ClusterManagerActorConfig)
         }
         result.map(CMCommandResults.apply) pipeTo sender()
 
+      case CMManualPartitionAssignments(assignments) =>
```
Comment (patelh) [inline app/kafka/manager/ClusterManagerActor.scala]: Can we add a test to exercise manual assignments?

Comment (terciodemelo) [inline app/kafka/manager/ClusterManagerActor.scala]: Done

File: app/kafka/manager/ClusterManagerActor.scala (hunk 7 above)
Comment (terciodemelo) [inline app/kafka/manager/ClusterManagerActor.scala]: Done

**Comment by patelh** [comment]
lgtm

**Comment by yazgoo** [comment]
Thanks @terciodemelo I could not get it to work. 
When clicking on Save Partition Assignment, nothing happens.
Also, I think it would be awesome to be able to export the reassignment JSON.

**Comment by terciodemelo** [comment]
@yazgoo After saving the partition assignment you must run them in the topics list view. I also think that it would be great some kind of assignment exporting feature - I can imagine a couple of uses with cluster visualization -  but it isn't possible for now, since the feature was temporarily removed from master.

**Pull request merged**